javascript/ql/lib/semmle/javascript/Routing.qll

```diff
@@ -0,0 +1,985 @@
+   *       - `r.use()`
+   *         - `auth()`
+   *   - `app.get(...)`
+   *     - `'/'`
+   *     - `handleRequest`
+   */
+  private class ImpliedRouteSetup extends RouteSetup::Base, DataFlow::InvokeNode {
+    Router::Range router;
+    Function target;
+
+    ImpliedRouteSetup() {
+      FlowSteps::calls(this, target) and
+      routerIsLiveInContainer(router, target) and
+      routerIsLiveInContainer(router, getContainer())
+    }
+
+    override Routing::Node getChild(int n) { result = MkRouter(router, target) and n = 0 }
+
+    override predicate isInstalledAt(Router::Range r, ControlFlowNode cfgNode) {
+      r = router and
+      cfgNode = getEnclosingExpr()
+    }
+  }
+
+  /**
+   * A function that handles an incoming request.
+   */
+  class RouteHandler extends Node {
+    DataFlow::FunctionNode function;
+
+    RouteHandler() { this = MkValueNode(function) }
+
+    /**
+     * Gets the `i`th parameter of this route handler.
+     *
+     * This is equivalent to `getParameter(i)` but returns a `RouteHandlerInput`.
+     *
+     * To find all references to this parameter, use `getInput(n).ref()`.
+     */
+    final RouteHandlerInput getInput(int n) { result = function.getParameter(n) }
+
+    /**
+     * Gets a parameter of this route handler.
+     *
+     * This is equivalent to `getAParameter()` but returns a `RouteHandlerInput`.
+     *
+     * To find all references to a parameter, use `getAnInput().ref()`.
+     */
+    final RouteHandlerInput getAnInput() { result = function.getAParameter() }
+
+    /** Gets the function implementing this route handler. */
+    DataFlow::FunctionNode getFunction() { result = function }
+
+    /**
+     * Gets a call that delegates the incoming request to the next route handler in the stack,
+     * usually a call of the form `next()`.
+     *
+     * By default, any 0-argument invocation of one of the route handler's parameters
+     * is considered a continuation invocation, since the other parameters (request and response)
+     * will generally not be invoked as a function. Framework models may override this method
+     * if the default behavior is inadequate for that framework.
+     */
+    DataFlow::CallNode getAContinuationInvocation() {
+      result = getAnInput().ref().getAnInvocation() and
+      result.getNumArgument() = 0
+      or
+      result.(DataFlow::MethodCallNode).getMethodName() = "then" and
+      result.getArgument(0) = getAnInput().ref().getALocalUse()
+    }
+  }
+
+  /**
+   * Gets the `RouteHandler` node corresponding to the given function.
+   *
+   * This has the same result as `getNode(function)` but is declared with a different return type.
+   */
+  RouteHandler getRouteHandler(DataFlow::FunctionNode function) { result.getFunction() = function }
+
+  /**
+   * A parameter to a route handler function.
+   */
+  class RouteHandlerInput extends DataFlow::ParameterNode {
+    RouteHandlerInput() { this = any(RouteHandler h).getFunction().getAParameter() }
```

Comment:
```suggestion
    RouteHandler handler;

    RouteHandlerInput() { this = handler.getFunction().getAParameter() }
```

You use the handler in two of the member predicates, so I think it's better in a field. 
(I've added suggestions to use the new field in the relevant member predicates).